components/region_cache_memory_engine/src/background.rs

```diff
@@ -547,35 +569,73 @@ impl Runnable for BackgroundRunner {
                         }
 
                         if canceled {
+                            info!(
+                                "snapshot load canceled due to memory reaching soft limit";
+                                "range" => ?range,
+                            );
                             core.on_snapshot_load_canceled(range);
                             continue;
                         }
-                        let start = Instant::now();
-                        for &cf in DATA_CFS {
-                            let guard = &epoch::pin();
-                            let handle = skiplist_engine.cf_handle(cf);
-                            match snap.iterator_opt(cf, iter_opt.clone()) {
-                                Ok(mut iter) => {
-                                    iter.seek_to_first().unwrap();
-                                    while iter.valid().unwrap() {
-                                        // use 0 sequence number here as the kv is clearly
-                                        // visible
-                                        let mut encoded_key =
-                                            encode_key(iter.key(), 0, ValueType::Value);
-                                        let mut val =
-                                            InternalBytes::from_vec(iter.value().to_vec());
-                                        encoded_key
-                                            .set_memory_controller(core.memory_controller.clone());
-                                        val.set_memory_controller(core.memory_controller.clone());
-                                        handle.insert(encoded_key, val, guard);
-                                        iter.next().unwrap();
+
+                        let snapshot_load = || -> bool {
+                            for &cf in DATA_CFS {
+                                let handle = skiplist_engine.cf_handle(cf);
+                                let guard = &epoch::pin();
+                                match snap.iterator_opt(cf, iter_opt.clone()) {
+                                    Ok(mut iter) => {
+                                        iter.seek_to_first().unwrap();
+                                        while iter.valid().unwrap() {
+                                            // use 0 sequence number here as the kv is clearly
```

Comment:
what if there's concurrent reading snapshot with smaller seq number. I think it's safer to use the current snapshot's seq number instead.